### ISSUE TITLE
Expose UI audio assignments to Unity editor

### ISSUE THREAD

**Issue by jefforkin** [giantotter/giantotter_public#338]
In order to allow Alex to easily place/replace audio clips for UI interactions, we would like to extend the audio manager to include a data structure for associating clips with UI events.  Making this data public will expose it to the Unity editor.

**Comment by grahamboree** [comment]
Some thoughts about this task:

We don't have UI sounds hooked up yet at all, so we'd need to do that before we can have Alex iterate on them.

While we could expose a bunch of sounds through the AudioManager, though we'd have to do a lot of additional work to actually use those sounds.  We'd need to create a component that references a specific sound in the audio manager, (through a string -> audioClip dictionary or something) add that new component to every button, add another onClick event to that button that references a public click handler function in this new component, and have that component acutally play the sound.  This is probably more work than is necessary.  To be clear, I am **not** advocating that we should do this.

Unity's way of playing UI sounds is the following: http://answers.unity3d.com/answers/858488/view.html

If our goal is to allow Alex to iterate quickly, we can achieve this without having to modify any code.  We can attach placeholder sounds to button presses and UI events using the process described in the above link.  We can place all these sounds in an `Assets/Sounds/UI` folder, and name them descriptive things like "button_press.wav".  Alex can then go and replace these sounds in that folder on disk, without doing anything with asset references through the editor.  When she wants to test them, she can use the editor to play the game.  

This gives us a few significant benefits:
* No need to modify code.
* Any sounds that aren't referenced aren't included in builds.  
* Alex doesn't need to mess with sound references on the AudioManager in the MainMenu scene, and worry about accidentally breaking something.
* Because we're using Unity's asset referencing system, we can change the file type and/or name freely while preserving the links between it and where it needs to be used.
* **Coolest of all** she can run the game, alt-tab and change a sound while the game is running, alt-tab back and Unity will notice the asset changed, and play the new version without having to restart the game.

**Comment by jefforkin** [comment]
Thanks Graham. That's definitely helpful info, I did not know that there was support for playing audio directly in the 4.6 UI stuff.  I agree with the benefits that you've laid out.  But in the spirit of thinking through all options, there are a couple issues I see with that system:
1) Alex has to rename the audio files to match our placeholders, and manually track which files she is actually using outside of Unity.  Obviously she can do this with a spreadsheet, but it makes it a bit more difficult to look at the assets and figure out which sounds (or versions of sounds) we are using.  IMHO an ideal system would allow filenames to be preserved, and abstract the association of a filename with a UI event.
2) We will have some UI-related events that are not really Unity UI events, like when the timer has 10 seconds left, and we want to play an alert.  So, those cases will still require some code, but could still follow the plan you've layed out where engineers hook up placeholder sounds with intuitive names.

Regarding "Any sounds that aren't referenced aren't included in builds" and changing the file type -- seems like these hold with either solution, because I was envisioning using the editor's support to browse for audio clips.

I do agree that it would be cool to be able to swap sounds without needing to do a build at all.  That is a big benefit.  And not having to write new code to hook up most UI audio is also a big win.  I'd just like to talk to Alex about how she has worked with other projects before making a final decision.

**Comment by grahamboree** [comment]
Hey Dave,
  Just talked with Jeff about this, and Dan and I had a call last week discussing it as well.  We think that all UI buttons should use a the same "default button press" sound, except where noted on the spreadsheet.  Where there are specific UI button sounds on the spreadsheet, these buttons should use their specific sounds, rather than the default button press sound.  
  This approach gives us the most flexibility.  If we add buttons in the future, they'll use the default sound.  If we want to change this for all buttons, it's in one place.  If we want to add more specific sounds for other buttons as well, we can set them specifically in their prefabs.  
  The one exception to this is the settings button sound in the spreadsheet.  We think it's probably a bit overkill to have a specific button sound for the settings button, so just ignore it.  The settings button should use the default button press sound.  
  To be clear, this shouldn't involve any code changes, unless there are instances where oscript is creating buttons that we want to have non-default sounds.  All other sounds can be hooked up through the button click event in its prefab.  Feel free to ping me if you have any questions.

**Comment by jefforkin** [comment]
no longer relevant
